src/background/deinflect.ts

```diff
@@ -508,6 +508,16 @@ export function deinflect(word: string): CandidateWord[] {
           continue;
         }
 
+        // Continue if the rule introduces a duplicate in the reason chain,
+        // as it wouldn't make sense grammatically.
+        if (
+          thisCandidate.reasonChains.some((rc) =>
+            rc.some((r) => rule.reasons.includes(r))
+          )
+        ) {
```

Comment:
I might be overthinking this, but this feels like an O(n^2) kind of algorithm (although, to be fair, `rule.reasons` is most often only one a single element).

Would it be slightly more efficient to first convert `rule.reasons` into a hash set like so:

```suggestion
        const ruleReasons = new Set(rule.reasons);
        if (thisCandidate.reasonChains.flat().some((r) => ruleReasons.has(r))) {
```

?

Sorry for not picking this up the first time I looked at this PR.

Reply:
I'm not sure, but given that `rule.reasons` is mostly just a single element and 3 elements at maximum, I think the overhead of allocating and initializing the Set might be more costly than just doing the linear search.

Would it make sense to try to benchmark both approaches?

Reply:
```ts
bench(
  'duplicates search',
  () => {
    deinflect('踊らせられとけたくなかった');
  },
  { time: 20000 }
);
```
This benchmark produced the following results:

| name | hz | min | max | mean | p75 | p99 |   p995 |   p999  |   rme  | samples |
| ---- | ---- | ---- | ---- | ---- | ---- | ---- | ---- | ---- | ---- | ---- |
| duplicates search (includes / no flat) | 36,024.92 | 0.0240 | 0.8123 | 0.0278 | 0.0260 | 0.1336 | 0.1565 | 0.1947  |±0.13% |  720499 |
| duplicates search (includes / flat) | 35,209.22|  0.0248 | 0.8148 | 0.0284 | 0.0265 | 0.1327  |0.1519  |0.1891 | ±0.12%|   704185|
| duplicates search (set / no flat) |34,604.29 | 0.0252|  1.0279 | 0.0289 | 0.0270|  0.1320  |0.1485  |0.1817  |±0.12%  | 692086|
| duplicates search (set / flat)| 33,145.89 | 0.0265 | 1.2632 | 0.0302 | 0.0282|  0.1368  |0.1528 | 0.1877|  ±0.12%   |662918|

Reply:
Wow, thanks for doing that! It's not really about benchmarking but about algorithmic complexity so that we don't get tripped up if reasons ever ends up being extended or we get some pathological input. I don't think this is ever going to be on the hot path since the I/O involved in querying IndexedDB is almost certainly going to dwarf any optimizations we make here.

(And the `flat()` part is just about readability since it took me a while to parse the nested `some()` calls.)

Reply:
Ah, I see. So considering this, we should definitely go with the hash set then?

Reply:
It's not strictly necessary, so if you're not convinced we can keep it as it is. I just get nervous when I see loops within loops within loops like this 😅

Reply:
No, I'm already convinced. Also, as you are the one carrying the responsibility for the whole code base, I would always opt for what you say is the best approach :)